src/array.cr

```diff
@@ -1768,16 +1775,30 @@ class Array(T)
   end
 
   private def double_capacity
-    resize_to_capacity(@capacity == 0 ? 3 : (@capacity * 2))
+    if @capacity == 0
+      capacity = 2
+    elsif @capacity < 3
+      capacity = 4
+    else
+      capacity = Math.pw2ceil(@capacity + @capacity/2 + 1)
+    end
+    resize_to_capacity(capacity)
   end
 
   private def resize_to_capacity(capacity)
-    @capacity = capacity
-    if @buffer
-      @buffer = @buffer.realloc(@capacity)
-    else
-      @buffer = Pointer(T).malloc(@capacity)
+    if @buffer != @allocation
+      @capacity += @buffer - @allocation
+      @allocation.move_from(@buffer, @size)
+    end
+    if @capacity != capacity
```

Comment:
I think `@capacity < capacity` will be better. Because we can reuse what we already allocated.

Reply:
Usually queue has sustained size, and then `!=` will work adequately.

`resize_to_capacity` is called not only from `double_capacity`. May be, there should be flag "increase", and then `increase ? @capacity < capacity : @capacity != capacity` .

Reply:
> Usually queue has sustained size, and then != will work adequately.

we already have `@size` for it.

> `resize_to_capacity` is called not only from `double_capacity`. May be, there should be flag "increase", and then `increase ? @capacity < capacity : @capacity != capacity` .

Yes, but they are requesting more memory than `@capacity` without adding `@buffer - @allocation`. Therefore, `@capacity > capacity` is possible. Here is example.

```Crystal
a = [0, 1, 2, 3, 4]
5.times { a.shift }
a.concat([0, 1, 2])
```

Reply:
And what? It certainly should not forbid decrease of allocation size. Pleade, suggest something that will allow array to shrink its allocation. Ie differentiate situations where it should not shrink, and when it could shrink.

Reply:
Calling `realloc` to adjust size is slow, even though it is not too frequent. I would suggest
give a method to shrink size rather than automatic shrink.

Reply:
If we are going to "user should do it manually", then there is no sense in this PR at all. This PR is about adding "magic", and not "adding manual methods". User can use Deque, organaize priority queue, etc instead of using Array as a queue.

What about reallocing if `@capacity < capacitu || @capacity > 8 && @capacity / 4 > capacity` ?

And considering your example with `5.times { a.shift }; a.concat([1,2,3])` - it is pattern this PR aimed to fix. This PR is only about `push + shift` pattern. Improving for "any" pattern (for example, for "sorted insert" from #5153) should be next step.

It is possible to increase complexity of this PR and cover "sorted insert", "shift + concat", and all other combinations. But for this should decision from major contributors about future of this PR. It is not obvious if unification of Array and Deque is necessary at all (or at least, attractive).